auth0/src/main/java/com/auth0/android/provider/OAuthManager.java

```diff
@@ -0,0 +1,295 @@
+        String sentState = parameters.get(KEY_STATE);
+        if (receivedState != null && sentState != null && !receivedState.equals(sentState)) {
+            Log.e(TAG, String.format("Received state doesn't match. Received %s but expected %s", receivedState, sentState));
+            throw new AuthenticationException(ERROR_VALUE_ACCESS_DENIED, "The received state is invalid. Try again.");
+        }
+    }
+
+    private void checkNonceValue(String receivedNonce, String idToken) throws AuthenticationException {
+        String sentNonce = parameters.get(KEY_NONCE);
+        if (parameters.get(KEY_RESPONSE_TYPE).contains(RESPONSE_TYPE_ID_TOKEN) && (idToken == null || !hasValidNonce(sentNonce, idToken))) {
+            Log.e(TAG, String.format("Received nonce doesn't match. Received %s but expected %s", receivedNonce, sentNonce));
+            throw new AuthenticationException(ERROR_VALUE_ACCESS_DENIED, "The received nonce is invalid. Try again.");
+        }
+    }
+
+    private Uri buildAuthorizeUri() {
+        Uri authorizeUri = Uri.parse(account.getAuthorizeUrl());
+        Uri.Builder builder = authorizeUri.buildUpon();
+        for (Map.Entry<String, String> entry : parameters.entrySet()) {
+            builder.appendQueryParameter(entry.getKey(), entry.getValue());
+        }
+        Uri uri = builder.build();
+        logDebug("Using the following AuthorizeURI: " + uri.toString());
+        return uri;
+    }
+
+    private void addPKCEParameters(Map<String, String> parameters, String redirectUri) {
+        if (!shouldUsePKCE()) {
+            return;
+        }
+        try {
+            pkce = createPKCE(redirectUri);
+            String codeChallenge = pkce.getCodeChallenge();
+            parameters.put(KEY_CODE_CHALLENGE, codeChallenge);
+            parameters.put(KEY_CODE_CHALLENGE_METHOD, METHOD_SHA_256);
+            Log.v(TAG, "Using PKCE authentication flow");
+        } catch (IllegalStateException e) {
+            Log.e(TAG, "Some algorithms aren't available on this device and PKCE can't be used. Defaulting to token response_type.", e);
+        }
+    }
+
+    private void addValidationParameters(Map<String, String> parameters) {
+        String state = getRandomString(parameters.get(KEY_STATE));
+        parameters.put(KEY_STATE, state);
+
+        if (parameters.get(KEY_RESPONSE_TYPE).contains(RESPONSE_TYPE_ID_TOKEN)) {
+            String nonce = getRandomString(parameters.get(KEY_NONCE));
+            parameters.put(KEY_NONCE, nonce);
+        }
+    }
+
+    private void addClientParameters(Map<String, String> parameters, String redirectUri) {
+        if (account.getTelemetry() != null) {
+            parameters.put(KEY_TELEMETRY, account.getTelemetry().getValue());
+        }
+        parameters.put(KEY_CLIENT_ID, account.getClientId());
+        parameters.put(KEY_REDIRECT_URI, redirectUri);
+    }
+
+    private PKCE createPKCE(String redirectUri) {
+        if (pkce != null) {
```

Comment:
It's just for testing purposes. https://github.com/auth0/Auth0.Android/pull/61/files/c4947e1f4f2cc68686a1479e7f3ec6d8b97f7696#diff-fba4899390d3e4efe09155da3e2a8396R80